### PR TITLE
feat: WIP add a discriminator to the change event source

### PR DESCRIPTION
**Pull request by wallyworld**
Events emitted from the change log table contain the id of the entity/row that has changed. However, sometimes filtering needs to be done based on only wanting events for a subset of the rows. 

Consider the case where we want to watch for changes to units belonging to a specific application. We could receive the event and go and look up the application uuid from the unit id in the event. Not only does this require a db access, it also totally fails when the event is for a unit deletion, because the row no longer exists.

This PR adds the possibility to include an additional column value - the discriminator - with the change event. The change_log table gets a new discriminator column and the trigger generation functions are able to optionally take an additional column to use to populate it. As part of the work, the trigger function args are re-ordered so the namespace goes first, then the event id column, then the optional discriminator column. It's nicer this way.

The expectation is that the discriminator column would be for an invariant value of the entity is is akin to a "unique key" of sorts. eg for units, that would be application_uuid. This then allows simple in memory filtering to be done in the watcher mapper function.

This is needed to be able to implement a lifecycle watcher with the same semantic as we had with mongo. In the mongo case, the doc id was not opaque - it contained enough info to filter on.

The first use case illustrating how this would be used is the units watcher for a given application - it's included here for reference but will landed as a separate PR. The integration test for this watcher illustrates the correct operation of the infrastructure. Before these changes, the event for a unit from a different app would incorrectly be emitted.

## QA steps

Just unit tests for now.

## Links

**Jira card:** JUJU-

### CODE REVIEW

File: internal/changestream/stream/stream.go [hunk 1]
```diff
@@ -456,7 +456,7 @@ const (
 	// If the namespace is later deleted, you'll no longer locate that during
 	// a select.
 	selectQuery = `
-SELECT MAX(c.id), c.edit_type_id, n.namespace, changed, created_at
+SELECT MAX(c.id), c.edit_type_id, n.namespace, changed, discriminator, created_at
 	FROM change_log c
 		JOIN change_log_edit_type t ON c.edit_type_id = t.id
 		JOIN change_log_namespace n ON c.namespace_id = n.id
```
Comment (SimonRichardson) [inline internal/changestream/stream/stream.go:462]: All events are coalesced at the database level via the use of the GROUP BY in this query. If the discriminator is not the same for the namespace and the changed value throughout the usage of the event then the discriminator will not be preserved (FWW). In other words, the discriminator MUST be a constant.

File: generate/triggergen/main.go [hunk 2]
```diff
@@ -302,22 +302,37 @@ package {{.Package}}
...
+		panic("too many table columns supplied")
+	}
+	changeLogColumns := []string{"changed"}
+	newColumnValues := "NEW." + changeColumnNames[0]
+	oldColumnValues := "OLD." + changeColumnNames[0]
+	if len(changeColumnNames) > 1 {
+		changeLogColumns = append(changeLogColumns, "discriminator")
+		newColumnValues += ", NEW." + changeColumnNames[1]
+		oldColumnValues += ", OLD." + changeColumnNames[1]
+	}
```
Comment (SimonRichardson) [inline generate/triggergen/main.go:327]: Don't overload this function. This complicates all trigger code. We can push this into the generation code at _compile_ time and not _runtime_.

File: internal/changestream/stream/stream.go (hunk 1 above)
Comment (wallyworld) [reply internal/changestream/stream/stream.go:462]: Yes, it is. It is invariant. eg units have an application_uuid and it never changes once set.

File: generate/triggergen/main.go (hunk 2 above)
Comment (wallyworld) [reply generate/triggergen/main.go]: Done

**Review by SimonRichardson — COMMENTED** [review]
I wouldn't filter the discriminator at the mapper level, you can do that at the changestream subscription level. The ChangeEvent has the discriminator there, you can add a `FilterNamespaceWithDiscriminator`:

https://github.com/juju/juju/blob/c65cd1593acc23189ea773bef239ff9a5c4f1632/core/changestream/subscription.go#L60-L66

This would prevent additional events being emitted through the system, only to be turned away at the mapper level.

**Review by manadart — COMMENTED** [review]
This is a slippery slope that I'd rather not jump on right now.

For the use-case you have, I feel you can do it in similar fashion to how the machines watcher is done (I saw something today that I need to fix, so emphasis on _similar_).

- Have an initial statement that populates based on your required watcher scope.
- For each event, query UUID, name and app, so all additions and changes build up the context. You can use a mapper for this.
- If you get add/delete masked by delete in a single term, you just don't find it and never add it.
- If you get a delete and you don't find it in the query, and you've got it in your cache, just delete it.
- For the last, you had it in the cache, so you can emit the unit name or app or whatever.

I don't think we should be concerned about read magnitude at this point. @SimonRichardson  is going to do some query-only benchmarking next pulse to see if we can approach some threshold, but attempts to minimise query volume by design changes is something we should delay.

**Pull request closed without merging**